### ISSUE TITLE
TS 2564 error reported incorrectly for overridden fields in a child class

### ISSUE THREAD

**Issue by octogonz** [microsoft/TypeScript#46499]
# Bug Report
 

### 🔎 Search Terms

`override` `class` `2564`

### 🕗 Version & Regression Information

This happens in any recent compiler if `strictPropertyInitialization` is enabled.  For example `4.4.4` 
 

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.4#code/LAKAlgdgLgpgTgMwIYGMYAICSAhJBnGAeQAcowB7CPdAb1HXQA8AudCAVwFsAjeAblABfUKBQAbfNVwFa9dMXbcxYFOnKkKVVjnxENlPAJByFSlehQGocdiijk4ACnVkD26XtdUAlLJAMGKAALMDwAOhdNagBeNX0qIwZhEGTQSFhEVAxMAGEQsQATEi9qGEZYCALqHQJiqL8GAE9WDh5+IVEJPGo8sEL0Moqq9A8G+UVlVUi3LF7CuoMjEwnzSyprW3snaa1Z-KL4vF86fwC8dmJ4Z0PvRPRk1OMQMRgoC330WIgYAHd0OYKjhoTGYAEYADToZoAJnut1EBnILzCYnIAHNHCh9hFDmFGvCQEA)

### 💻 Code

```ts
interface IBaseOptions {
  x: number;
}

class Base {
  public options: IBaseOptions;

  public constructor(options: IBaseOptions) {
    this.options = options;
  }
}

interface IChildOptions extends IBaseOptions {
  y: number;
}
class Child extends Base {
  // 👇 "Property 'options' has no initializer and is not definitely assigned in the constructor."
  public options: IChildOptions; 

  public constructor(options: IChildOptions) {
    super(options);
  }
}
```
### 🙁 Actual behavior

The `"not definitely assigned in the constructor"` error is reported.

### 🙂 Expected behavior

No errors should be reported for this code sample.

### Remarks

The "not definitely assigned" message doesn't make sense for fields that are overridden in a child class.  If the base class did neglect to initialize the field, then this error would already be reported in the base class.    The child class should not be second-guessing that validation.

**Comment by fatcerberus** [comment]
Here's why TS is second-guessing you:
```ts
interface IBaseOptions {
  x: number;
}

class Base {
  public options: IBaseOptions;

  public constructor(options: IBaseOptions) {
    this.options = options;
  }
}

interface IChildOptions extends IBaseOptions {
  y: number;
}
class Child extends Base {
  // pretend there's no error here
  public options: IChildOptions; 

  public constructor() {
    super({ x: 812 });  // no error
  }
}

let foo = new Child();
console.log(foo.options.y);  // supposed to be number, but prints undefined
```

Yes, the base class initializes it if you don't, but then it's not guaranteed to be the right type.

**Comment by andrewbranch** [comment]
@fatcerberus’s answer is a convenient way for the compiler to save face, but the truth is that the definite assignment analysis has significant limitations. What you have is just an even more complicated version of https://www.typescriptlang.org/play?ts=4.4.4#code/JYOwLgpgTgZghgYwgAgJIGEAWwA2ATAeQAcxgB7EAZ2QG8BYAKGWQE8AuZEAVwFsAjaAG5GAX0YIccStSy48tRsyJc+OYAmRkS5KhwzZ8xUhUrCGi5MtXrkCE2ChcEYMlAAUW47rSzD2kwCUCkzMyGDYlAB0lBBgRjqUHv5UAWbMYuYhVmoaMXHJiZ4Jer6EBUH0IczhwFFFJsgAvJoFacgZGUA, which is safe, but not something we’ll recognize. It’s expected that you’ll sometimes have to use `public options!: IChildOptions` if you step outside of some very simple patterns.

**Comment by fatcerberus** [comment]
@andrewbranch For what it's worth, I wouldn't really expect the pattern in the OP to work, even if your example did.  Even assuming that static analysis could descend into function calls, from the type system point of view, `options` is being initialized with something of type `IBaseOptions`, *not* `IChildOptions`--so the error in this case is desirable because there's no (type-level) guarantee that `super()` gets called with the correct type.
